### ISSUE TITLE
Async result still applied even when value or condition changed

### ISSUE THREAD

**Issue by lostincomputer** [Knockout-Contrib/Knockout-Validation#549]
The result of async rule is still being applied even when value or condition has changed. This causes Knockout Validation to mark the observable as invalid even when should be valid.
The result of async rule should be ignored if the observable's value has changed or the rule's condition/onlyIf turned to false.

Timeline:
1. observable is set to an invalid value.
2. Knockout Validation runs the async rule. Lets call the run *async#1*. 
3. observable is set to a valid value.
4. Knockout Validation runs the async rule. Lets call the run *async#2*.
5. *async#2* finishes and calls callback(true)
6. *async#1* finishes and calls callback(false)
7. observables.isValid is false when it should be true!

Same problem occurs when the rule's condition/onlyIf is changed to false after the async rule is run but before the async rule finishes.

**Comment by ncortines** [comment]
Hi there,

I'm not sure, but I think I'm facing similar issue.

In my case the problem is that the last async validation fired does not cancel off the previous one, which yields an invalid validation state in some cases.

I've written a small example to demonstrate it:
http://jsfiddle.net/8LxL9u9v/9/

Just type "TT" and then, on quick sucession "TF" and "TT".

The final state should be valid, but instead it is marked as invalid because the last async validation takes less time to compute.

Thanks

**Comment by crissdev** [comment]
The condition is used only to determine if the validation should run, not to set the validation result. This can be overcome by watching `isValidating` property of the observable and prevent changes that would cause re-validation. Also, check the jsFiddle example bellow.

@ncortines I've modified the example to delay validation when you type in the input. The control is also changed to be read-only when validation is active. The example also makes use of `throttle` option to delay the validation.

http://jsfiddle.net/rs817sw8/

**Comment by lostincomputer** [comment]
Unfortunately, our application is required to not block the user when validation is running. We can't set the control to readonly or disable (or the users and BA will get mad).

**Comment by ncortines** [comment]
@crissdev , thanks for the example, it is a goof workaround.

However I have to agree with @lostincomputer . The current behavior does not seem to be correct.

**Comment by lostincomputer** [comment]
@crissdev I forgot to mention that our async validators do use debouncers (delay) but it gets defeated when the network or server is too slow. It's rare but it happens.

**Comment by crissdev** [comment]
Thanks for your input. Async validation can really be a pain sometimes. But we can try to make it better :smiley: 

Considering to give the last ran validation the final word on the result would imply to discard any previous callback results until the last pending operation completes. I hope I'll explain this well and provide some example tomorrow.
This could be achieved by storing a _token_ for each async validation and checking it when the operation completes. If the token was not created by the running operation then it should be discarded, until the last operation completes. This means that if 5 validations are running then only the last one should affect the validation state.

To summarize:

1. Observable changes 2 times causing 2 async validations (__op1__ and __op2__)
1.1 Each async validation call creates a token and stores it on the observable (e.g. __token__)
2. __op1__ completes but the __token__ is different :arrow_right: discard result
3. __op2__ completes and the __token__ is the same the operation had set :arrow_right: update validation state

Any words on this?

**Comment by ncortines** [comment]
In my opinion discarding the **op1** result while **op2** is still in progress is wasting valuable information, since it informs the user about the latest valid state.

I would make this **token** a variable which is incremented per each validation fired. When the callback is finally executed with the validation result, it should check whether the new state is valid by comparing the **token** with the **token** that successfully updated the **token** for the last time:

1. **op1** is fired with **token** 1
2. **op2** is fired with **token** 2
3. **op3** is fired with **token** 3
4. **op1** resolves and updates the state. At this point **op2** and **op3** have not resolved yet.
5. **op3** resolves and updates the state (3 > 1). At this point **op3** has not resolved yet.
6. **op2** resolves but does not update the state because it's token is not valid (2 < 3)

**Comment by lostincomputer** [comment]
Showing stale validation state can mislead the user though unless the UI has an 'is validating' indicator.

By the way, isValidating property doesn't support concurrent async operations. The easiest way to fix this is to add a counter that is incremented when an async operation begins and decremented when an async operation finishes.

isValidating can then be converted to a computed property:

isValidating = ko.computed(function() {
       return counter() > 0;
});

**Comment by ncortines** [comment]
Keeping the latest **valid** validation state until a new async validation resolves is more consistent with respect to the current implementation, where every resolved async operation updates the state. The only change here would be to discard those updates considered as **not valid**.

The additional change to discard the results from **all** the intermediate concurrent async validations (other than the last one to resolve) would a suppose a breaking change in the api for some users like me. On the other hand this could be accomplished by other external means, e.g. aborting an ajax request, etc.

Finally the solution proposed above, having a counter which is decremented with each operation resolved, does not guarantee that the final state update is performed by the last validation fired, which is my current problem.

**Comment by ncortines** [comment]
You are right. Intermediate results could be safely discarded. Valid unless proven invalid. I'm fine with the solution described above by @crissdev

**Comment by lostincomputer** [comment]
There are two reasons for isValidating to be updated. First: multiple async rules attached to the observable. Second: the async rule is still running when it is called again.

The counter will solve the first case by keeping track of the number of running async rules.

P.S. We don't attached multiple async rules in our observables but if other people will or does, modifying isValidating will be great.

**Comment by crissdev** [comment]
Thank you guys for your help on this.

Using some random string or a counter will both help in keeping the `isValidating` property consistent. The advantage of using a counter is it helps when debugging.

@lostincomputer There's no need to make `isValidating` a computed property because we might need to manually set that property in the future.
